### PR TITLE
fix: infinite recursion in mysql connector and improve mysql EE bridge tests

### PR DESCRIPTION
**Pull request by olcai**
This is a continuation of EMQX-7923 and implements further EE mysql bridge tests covering:

- prepared statements
- batching
- driver behavior with timeout failures

A bug was found while implementing tests regarding prepared statements. An infinite loop was triggered in the mysql connector when a query used a prepared statement key that was not among the defined prepared statements on start. We now check that the key is defined among the prepared statements before recursing. It seems that this bug was never
triggered in any production code.

Fixes EMQX-8414.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [X] Added tests for the changes
- [X] Changed lines covered in coverage report
- [x] Change log has been added to `changes/` dir
- [X] For internal contributor: there is a jira ticket to track this change

### CODE REVIEW

File: changes/v5.0.13-zh.md [hunk 1]
```diff
@@ -20,3 +20,5 @@
 
 - 修复共享订阅的 'sticky' 策略 [#9578](https://github.com/emqx/emqx/pull/9578)。
   在此修复前，使用 'sticky' 策略的订阅客户端可能在取消订阅后继续收到消息。
+
+- Add check to ensure that a given key is among the prepared statements on query in the mysql connector [#9571](https://github.com/emqx/emqx/pull/9571).
```
Comment (olcai) [inline changes/v5.0.13-zh.md:24]: Translation required

File: lib-ee/emqx_ee_bridge/test/emqx_ee_bridge_mysql_SUITE.erl [hunk 2]
```diff
@@ -392,3 +410,114 @@ t_write_failure(Config) ->
...
+    Timeout = 10,
+    ?check_trace(
+        emqx_common_test_helpers:with_failure(timeout, ProxyName, ProxyHost, ProxyPort, fun() ->
+            query_resource(Config, {send_message, SentData, [], Timeout})
+        end),
+        fun(Result, _Trace) ->
+            ?assertMatch({error, {resource_error, _}}, Result),
+            ok
+        end
+    ),
```
Comment (thalesmg) [inline lib-ee/emqx_ee_bridge/test/emqx_ee_bridge_mysql_SUITE.erl:432]: Minor nit: since the trace is not being used, we might just drop `?check_trace`.

```suggestion
    emqx_common_test_helpers:with_failure(timeout, ProxyName, ProxyHost, ProxyPort, fun() ->
        ?assertMatch({error, {resource_error, _}}, query_resource(Config, {send_message, SentData, [], Timeout}))
    end),
```

**Review by thalesmg — APPROVED** [review]

File: changes/v5.0.13-zh.md (hunk 1 above)
Comment (lafirest) [reply changes/v5.0.13-zh.md:24]: - 增强 mysql 查询流程，确保给定的查询语句在 mysql 连接器的预编译语句中 [#9571](https://github.com/emqx/emqx/pull/9571)。

Comment (olcai) [reply changes/v5.0.13-zh.md]: Thank you!

File: changes/v5.0.13-zh.md [hunk 3]
```diff
@@ -20,3 +20,5 @@
 
 - 修复共享订阅的 'sticky' 策略 [#9578](https://github.com/emqx/emqx/pull/9578)。
   在此修复前，使用 'sticky' 策略的订阅客户端可能在取消订阅后继续收到消息。
+
+- 增强 mysql 查询流程，确保给定的查询语句在 mysql 连接器的预编译语句中 [#9571](https://github.com/emqx/emqx/pull/9571)。
```
Comment (qzhuyan) [inline changes/v5.0.13-zh.md:24]: LGTM

**Review by thalesmg — APPROVED** [review]

File: apps/emqx_connector/src/emqx_connector_mysql.erl [hunk 4]
```diff
@@ -256,6 +260,12 @@ init_prepare(State = #{prepare_statement := Prepares, poolname := PoolName}) ->
             end
     end.
 
+maybe_prepare_sql(SQLOrKey, Prepares, PoolName) ->
+    case maps:is_key(SQLOrKey, Prepares) of
+        true -> prepare_sql(Prepares, PoolName);
+        false -> {error, prepared_statement_is_unprepared}
```
Comment (zmstone) [inline apps/emqx_connector/src/emqx_connector_mysql.erl:266]: ?
```suggestion
        false -> {error, prepared_statement_invalid}
```

File: apps/emqx_connector/src/emqx_connector_mysql.erl [hunk 5]
```diff
@@ -305,6 +315,8 @@ prepare_sql_to_conn(Conn, [{Key, SQL} | PrepareList]) when is_pid(Conn) ->
             ?SLOG(info, LogMeta#{result => success}),
             prepare_sql_to_conn(Conn, PrepareList);
         {error, Reason} ->
+            % FIXME: we should try to differ on transient failers and
```
Comment (zmstone) [inline apps/emqx_connector/src/emqx_connector_mysql.erl:318]: Possible to fix now?

File: apps/emqx_connector/src/emqx_connector_mysql.erl (hunk 4 above)
Comment (olcai) [reply apps/emqx_connector/src/emqx_connector_mysql.erl]: Yep, that's a better name.

File: apps/emqx_connector/src/emqx_connector_mysql.erl (hunk 5 above)
Comment (olcai) [reply apps/emqx_connector/src/emqx_connector_mysql.erl:318]: I took a shot at it but came to the conclusion that it might require a more thorough refactor. The code is hard to follow as-is and already has several layers of error coping mechanisms, and these should be unified. Created EMQX-8582 to track it.

**Review by zmstone — APPROVED** [review]

**Comment by zmstone** [comment]
the slim build fails will be fixed in another PR.
